File: code/primes.cpp
/** This code was at Google I/O 2012: 
Breaking the JavaScript Speed Limit with V8 
(http://www.youtube.com/watch?v=UJPdhx5zTaw) **/

class Primes {
	public:
		int getPrimeCount() const { return prime_count; }
		int getPrime(int i) const { return primes[i]; }
		void addPrime(int i) { primes[prime_count++] = i; }
		
		bool isDivisible(int i, int by) { return (i % by) == 0; }

		bool isPrimeDivisible(int candidate) {
			for (int i = 0; i < prime_count; i++) {
				if (isDivisible(candidate, primes[i])) return true;
			}
			return false;
		}
	private:
		volatile int prime_count;
		volatile int primes[25000];
};

int main() {
	Primes p;
	int c = 1;
	while (p.getPrimeCount() < 25000) {
		if (!p.isPrimeDivisible(c)) {
			p.addPrime(c);
		}
		c++;
	}
	printf("%d\n", p.getPrime(p.getPrimeCount() - 1))
}

File: code/bench/primes.cs
using System;

namespace Application
{
	public class Primes
	{
		public int prime_count;
		public int[] primes = new int[5000];

		public int getPrimeCount () { return this.prime_count; }
		public int getPrime (int i) { return this.primes[i]; }
		public void addPrime (int p) { this.primes[this.prime_count++] = p; }

		public bool isDivisible(int i, int by) { return (i % by) == 0; }

		public bool isPrimeDivisible(int candidate) {
			for (int i = 1; i < this.prime_count; ++i) {
				if (this.isDivisible(candidate, this.primes[i])) return true;
			}
			return false;
		}
	}

	public static class Programm {
		public static void Main() {
			Primes p = new Primes();
			int c = 1;
			while (p.getPrimeCount() < 5000) {
				if (!p.isPrimeDivisible(c)) {
					p.addPrime(c);
				}
				c++;
			}
			System.Console.WriteLine(p.getPrime(p.getPrimeCount() - 1));
		}
	}

}

